File: frontend/src-tauri/src/recall/chunker.rs
//! Split a meeting's ordered transcript segments into overlapping windows suitable for
//! embedding + retrieval. Each chunk carries the recording-relative time span and a
//! display timestamp so retrieved chunks can cite a moment.

use crate::database::models::Transcript;

/// Target chunk size in characters (~500 tokens at ~4 chars/token). Keeps chunks well
/// inside the embedder's context while staying large enough to hold a coherent thought.
const TARGET_CHARS: usize = 2_000;
/// Segments carried into the next chunk so a fact spanning a boundary stays retrievable.
const OVERLAP_SEGMENTS: usize = 1;

pub struct ChunkDraft {
    pub chunk_index: i64,
    pub text: String,
    pub start_time: Option<f64>,
    pub end_time: Option<f64>,
    pub timestamp_label: Option<String>,
    pub token_estimate: i64,
}

fn build_chunk(index: i64, segments: &[&Transcript]) -> Option<ChunkDraft> {
    let text = segments
        .iter()
        .map(|s| s.transcript.trim())
        .filter(|t| !t.is_empty())
        .collect::<Vec<_>>()
        .join(" ");
    if text.trim().is_empty() {
        return None;
    }
    let start_time = segments.iter().find_map(|s| s.audio_start_time);
    let end_time = segments.iter().rev().find_map(|s| s.audio_end_time);
    let timestamp_label = segments
        .iter()
        .find(|s| !s.timestamp.trim().is_empty())
        .map(|s| s.timestamp.trim().to_string());
    let token_estimate = (text.chars().count() / 4) as i64;
    Some(ChunkDraft {
        chunk_index: index,
        text,
        start_time,
        end_time,
        timestamp_label,
        token_estimate,
    })
}

/// Chunk transcript segments (expected in chronological order) into overlapping windows.
pub fn chunk_transcripts(segments: &[Transcript]) -> Vec<ChunkDraft> {
    let mut chunks = Vec::new();
    let mut current: Vec<&Transcript> = Vec::new();
    let mut current_chars = 0usize;
    let mut index = 0i64;

    for segment in segments {
        let text = segment.transcript.trim();
        if text.is_empty() {
            continue;
        }
        current.push(segment);
        current_chars += text.chars().count();

        if current_chars >= TARGET_CHARS {
            if let Some(draft) = build_chunk(index, &current) {
                chunks.push(draft);
                index += 1;
            }
            let keep_from = current.len().saturating_sub(OVERLAP_SEGMENTS);
            current = current[keep_from..].to_vec();
            current_chars = current
                .iter()
                .map(|s| s.transcript.trim().chars().count())
                .sum();
        }
    }

    // Flush the tail, unless it's only the carried-over overlap (already covered).
    let is_pure_overlap = !chunks.is_empty() && current.len() <= OVERLAP_SEGMENTS;
    if !current.is_empty() && !is_pure_overlap {
        if let Some(draft) = build_chunk(index, &current) {
            chunks.push(draft);
        }
    }

    chunks
}

#[cfg(test)]
mod tests {
    use super::*;

    fn seg(id: &str, text: &str, start: f64) -> Transcript {
        Transcript {
            id: id.to_string(),
            meeting_id: "m1".to_string(),
            transcript: text.to_string(),
            timestamp: format!("{:02}:{:02}", (start as i64) / 60, (start as i64) % 60),
            summary: None,
            action_items: None,
            key_points: None,
            audio_start_time: Some(start),
            audio_end_time: Some(start + 5.0),
            duration: Some(5.0),
            speaker_id: None,
        }
    }

    #[test]
    fn short_transcript_makes_one_chunk_with_time_span() {
        let segments = vec![seg("a", "hello world", 0.0), seg("b", "second line", 6.0)];
        let chunks = chunk_transcripts(&segments);
        assert_eq!(chunks.len(), 1);
        assert_eq!(chunks[0].start_time, Some(0.0));
        assert_eq!(chunks[0].end_time, Some(11.0));
        assert_eq!(chunks[0].timestamp_label.as_deref(), Some("00:00"));
        assert!(chunks[0].text.contains("hello world"));
        assert!(chunks[0].text.contains("second line"));
    }

    #[test]
    fn long_transcript_splits_into_multiple_chunks() {
        let big = "word ".repeat(500); // ~2500 chars
        let segments = vec![seg("a", &big, 0.0), seg("b", &big, 60.0), seg("c", "tail", 120.0)];
        let chunks = chunk_transcripts(&segments);
        assert!(chunks.len() >= 2, "expected multiple chunks, got {}", chunks.len());
        // Indices are sequential from 0.
        for (expected, chunk) in chunks.iter().enumerate() {
            assert_eq!(chunk.chunk_index, expected as i64);
        }
    }

    #[test]
    fn empty_input_yields_no_chunks() {
        assert!(chunk_transcripts(&[]).is_empty());
    }
}

File: AriKit/Sources/AriKit/Recall/Chunking/Chunker.swift
//
//  Chunker.swift — transcript chunking for the recall index (plan §7, ← chunker.rs).
//
//  Split a meeting's ordered transcript segments into overlapping windows suitable for embedding
//  + retrieval. Each chunk carries the recording-relative time span and a display timestamp so a
//  retrieved chunk can cite a moment. Consumes the ported `Models.Transcript`; char counts are in
//  Unicode scalars to match Rust `chars().count()`.
//
import Foundation

/// A drafted chunk (← `ChunkDraft`, chunker.rs:13) — the pre-persistence value the indexer will
/// embed + store. `chunkIndex` is sequential from 0 within a meeting.
public struct ChunkDraft: Codable, Hashable, Sendable {
    public var chunkIndex: Int
    public var text: String
    public var startTime: Double?
    public var endTime: Double?
    public var timestampLabel: String?
    public var tokenEstimate: Int

    public init(
        chunkIndex: Int,
        text: String,
        startTime: Double? = nil,
        endTime: Double? = nil,
        timestampLabel: String? = nil,
        tokenEstimate: Int
    ) {
        self.chunkIndex = chunkIndex
        self.text = text
        self.startTime = startTime
        self.endTime = endTime
        self.timestampLabel = timestampLabel
        self.tokenEstimate = tokenEstimate
    }
}

extension Recall {
    /// Target chunk size in scalars (~500 tokens at ~4 chars/token) (← `TARGET_CHARS`).
    private static let targetChars = 2000
    /// Segments carried into the next chunk so a boundary-spanning fact stays retrievable
    /// (← `OVERLAP_SEGMENTS`).
    private static let overlapSegments = 1

    /// Chunk transcript segments (expected chronological) into overlapping windows
    /// (← `chunk_transcripts`).
    public static func chunkTranscripts(_ segments: [Transcript]) -> [ChunkDraft] {
        var chunks: [ChunkDraft] = []
        var current: [Transcript] = []
        var currentChars = 0
        var index = 0

        for segment in segments {
            let text = segment.transcript.trimmingCharacters(in: .whitespacesAndNewlines)
            if text.isEmpty {
                continue
            }
            current.append(segment)
            currentChars += text.unicodeScalars.count

            if currentChars >= targetChars {
                if let draft = buildChunk(index: index, segments: current) {
                    chunks.append(draft)
                    index += 1
                }
                let keepFrom = current.count - min(current.count, overlapSegments)
                current = Array(current[keepFrom...])
                currentChars = current.reduce(0) {
                    $0 + $1.transcript.trimmingCharacters(in: .whitespacesAndNewlines).unicodeScalars.count
                }
            }
        }

        // Flush the tail, unless it is only the carried-over overlap (already covered).
        let isPureOverlap = !chunks.isEmpty && current.count <= overlapSegments
        if !current.isEmpty, !isPureOverlap {
            if let draft = buildChunk(index: index, segments: current) {
                chunks.append(draft)
            }
        }

        return chunks
    }

    /// Chunk a meeting's generated summary body into overlapping windows, mirroring
    /// `chunkTranscripts`'s target-size/overlap behavior but over one plain document rather than N
    /// ordered transcript segments (plan `ask-meetings-tools-and-cards.md` §3.2). No timestamp is
    /// ever attached — a summary chunk has no meaningful recording-relative time span, and this
    /// function stays summary-agnostic about WHY it's being called: it is the caller (`Indexer`)
    /// that tags the resulting drafts with `sourceKind == .summary`, not `Chunker` itself.
    public static func chunkSummary(_ bodyMarkdown: String) -> [ChunkDraft] {
        let paragraphs = bodyMarkdown
            .components(separatedBy: .newlines)
            .map { $0.trimmingCharacters(in: .whitespaces) }
            .filter { !$0.isEmpty }
        guard !paragraphs.isEmpty else { return [] }

        var chunks: [ChunkDraft] = []
        var current: [String] = []
        var currentChars = 0
        var index = 0

        func flush() {
            let text = current.joined(separator: "\n")
            if !text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
                chunks.append(ChunkDraft(
                    chunkIndex: index,
                    text: text,
                    tokenEstimate: text.unicodeScalars.count / 4
                ))
                index += 1
            }
        }

        for paragraph in paragraphs {
            current.append(paragraph)
            currentChars += paragraph.unicodeScalars.count
            if currentChars >= targetChars {
                flush()
                let keepFrom = current.count - min(current.count, overlapSegments)
                current = Array(current[keepFrom...])
                currentChars = current.reduce(0) { $0 + $1.unicodeScalars.count }
            }
        }

        let isPureOverlap = !chunks.isEmpty && current.count <= overlapSegments
        if !current.isEmpty, !isPureOverlap {
            flush()
        }

        return chunks
    }

    private static func buildChunk(index: Int, segments: [Transcript]) -> ChunkDraft? {
        let text = segments
            .map { $0.transcript.trimmingCharacters(in: .whitespacesAndNewlines) }
            .filter { !$0.isEmpty }
            .joined(separator: " ")
        if text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            return nil
        }
        let startTime = segments.lazy.compactMap(\.audioStartTime).first
        let endTime = segments.reversed().lazy.compactMap(\.audioEndTime).first
        let timestampLabel = segments
            .first { !$0.timestamp.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty }
            .map { $0.timestamp.trimmingCharacters(in: .whitespacesAndNewlines) }
        let tokenEstimate = text.unicodeScalars.count / 4
        return ChunkDraft(
            chunkIndex: index,
            text: text,
            startTime: startTime,
            endTime: endTime,
            timestampLabel: timestampLabel,
            tokenEstimate: tokenEstimate
        )
    }
}
